### ISSUE TITLE
bug: `redundant-import-alias` rule check for import path instead of package name

### ISSUE THREAD

**Issue by candiduslynx** [mgechev/revive#936]
**Describe the bug**

When checking `redundant-import-alias` the code will only look at the import path, whereas the proper way would've been to check the imported package name.
I've seen this with `import apiv2 "cloud.google.com/go/run/apiv2"` (the code is under `apiv2` directory, but the package name is `run`: https://pkg.go.dev/cloud.google.com/go/run/apiv2)

**To Reproduce**
1. Create a package in a dir named `redundant` with the following content (note that the package name must be different):
   ```go
   package relevant
   
   var Lovely = int(123)
   ```
3. Import it via `import redundant "path/to/redundant"`
4. Run check

**Expected behavior**
No issue is presented, as the alias is used to override the package name, not the path.

**Logs**
```
redundant-import-alias: Import alias "apiv2" is redundant (revive)
```
This is printed when importing:
```go
import apiv2 "cloud.google.com/go/run/apiv2"
```

**Comment by chavacava** [comment]
Hi @candiduslynx, thank you for reporting the issue.

**Comment by denisvmedia** [comment]
The only viable solution for that seems to be using

```
	import "golang.org/x/tools/go/packages"

	// ...

	// Load the package information
	pkgs, err := packages.Load(cfg, importPath)
	if err != nil {
		return "", fmt.Errorf("Error loading package: %w", err)
	}

	// Ensure that the package information is available
	if len(pkgs) == 0 || pkgs[0].Name == "" {
		return "", fmt.Errorf("Package information not found for import path: %s", importPath)
	}

	return pkgs[0].Name, nil // <-- here we'll return `run` for https://pkg.go.dev/cloud.google.com/go/run/apiv2
```

The problem of this solution is that package.Load uses `exec.Command` internally. And thus it's very slow on big code bases. For such cases I'd use `importas` linter, where you can strictly define an alias (or an absence of such). I'm not sure if `revive` has a built-in alternative.

**Comment by candiduslynx** [comment]
Another possible solution would be to load only package name when traversing the vendored imports, but this will also require an additional effort:
* How the vendor directory is defined based on the place where linter was executed from?
* Are the dependencies required to be vendored now? If now, this will be a major change to simple code scanning

**Comment by denisvmedia** [comment]
There are two ways of having the deps: `vendor` dir or in `$GOPATH/pkg/mod`. This would mean that the tool will have to understand, which mode is used in the project. While technically possible (first check for `vendor` dir, then - if not found - look in `$GOPATH/pkg/mod`), that would require the rule to go way beyond what other rules do. It will also mean that the rule will have to parse the first found non-test file using `"go/parser"` (this will slow down the processing as well, I guess). I'm not sure if it's a good idea.

**Comment by candiduslynx** [comment]
IMO this check just needs to be retired, as it's not usable with some major SDKs.

Or, at the very least, the package name should be determined by an updated process (it will still be a hack and won't work for the example in the issue description, where the package name doesn't match the directory name):
1. Split name parts
2. iterating from the end, check if the part is actually `v%d`
3. if it's `v%d`, go to the next one
4. if not - here's the (assumed) package name

**Comment by jpenna** [comment]
I just disabled the rule because it gave a false positive for:

```
qdrant "github.com/qdrant/go-client/qdrant"
```

**Comment by chavacava** [comment]
As mentioned earlier in this thread, accurately identifying a package in the general case often requires access to its mod file or source code, which could lead to potential performance issues and increased rule complexity. The current rule works well in the majority of cases. If it produces false positives in specific scenarios, it can be disabled [locally](https://github.com/mgechev/revive#comment-directives) or [globally](https://github.com/mgechev/revive#configuration).

**Comment by ccoVeille** [comment]
I find back this old PR with the issue #1662 raised recently that quoted this issue as relevant.

The conclusion here are still relevant I think. But they are old, and things evolve over the he time in Go ecosystem.

I'm telling this because I recently also saw the following PR

- https://github.com/golang/tools/pull/577

So apparently right now, there is already a way to get the module name via this

https://pkg.go.dev/golang.org/x/tools/go/analysis#Module

It might change nothing in term of performance being bad if we use it, but I wanted to drop the info just in case.
